pkg/network/protocols/kafka/protocol.go

```diff
@@ -0,0 +1,135 @@
+
+package kafka
+
+import (
+	"strings"
+	"unsafe"
+
+	manager "github.com/DataDog/ebpf-manager"
+	"github.com/cilium/ebpf"
+
+	"github.com/DataDog/datadog-agent/pkg/network/config"
+	"github.com/DataDog/datadog-agent/pkg/network/protocols"
+	"github.com/DataDog/datadog-agent/pkg/network/protocols/events"
+)
+
+type protocol struct {
+	cfg            *config.Config
+	telemetry      *Telemetry
+	statkeeper     *StatKeeper
+	eventsConsumer *events.Consumer
+}
+
+const (
+	eventStream                              = "kafka"
+	filterTailCall                           = "socket__kafka_filter"
+	dispatcherTailCAll                       = "socket__protocol_dispatcher_kafka"
+	protocolDispatcherClassificationPrograms = "dispatcher_classification_progs"
+	kafkaLastTCPSeqPerConnectionMap          = "kafka_last_tcp_seq_per_connection"
+)
+
+var Spec = protocols.ProtocolSpec{
+	Factory: newKafkaProtocol,
+	Maps: []*manager.Map{
+		{
+			Name: protocolDispatcherClassificationPrograms,
+		},
+		{
+			Name: kafkaLastTCPSeqPerConnectionMap,
+		},
+	},
+	TailCalls: []manager.TailCallRoute{
+		{
+			ProgArrayName: protocols.ProtocolDispatcherProgramsMap,
+			Key:           uint32(protocols.ProgramKafka),
+			ProbeIdentificationPair: manager.ProbeIdentificationPair{
+				EBPFFuncName: filterTailCall,
+			},
+		},
+		{
+			ProgArrayName: protocolDispatcherClassificationPrograms,
+			Key:           uint32(protocols.DispatcherKafkaProg),
+			ProbeIdentificationPair: manager.ProbeIdentificationPair{
+				EBPFFuncName: dispatcherTailCAll,
+			},
+		},
+	},
+}
+
+func newKafkaProtocol(cfg *config.Config) (protocols.Protocol, error) {
+	if !cfg.EnableKafkaMonitoring {
+		return nil, nil
+	}
+
+	telemetry := NewTelemetry()
+
+	return &protocol{
+		cfg:       cfg,
+		telemetry: telemetry,
+	}, nil
+}
+
+// ConfigureOptions add the necessary options for the http monitoring to work,
+// to be used by the manager. These are:
+// - Set the `http_in_flight` map size to the value of the `max_tracked_connection` configuration variable.
+//
+// We also configure the http event stream with the manager and its options.
+func (p *protocol) ConfigureOptions(mgr *manager.Manager, opts *manager.Options) {
+	events.Configure(eventStream, mgr, opts)
+	opts.MapSpecEditors[kafkaLastTCPSeqPerConnectionMap] = manager.MapSpecEditor{
+		Type:       ebpf.Hash,
+		MaxEntries: p.cfg.MaxTrackedConnections,
+		EditorFlag: manager.EditMaxEntries,
+	}
+}
+
+func (p *protocol) PreStart(mgr *manager.Manager) (err error) {
+	p.eventsConsumer, err = events.NewConsumer(
+		eventStream,
+		mgr,
+		p.processKafka,
+	)
+	if err != nil {
+		return
+	}
+
+	p.statkeeper = NewStatkeeper(p.cfg, p.telemetry)
+	p.eventsConsumer.Start()
+
+	return
+}
+
+func (p *protocol) PostStart(_ *manager.Manager) error {
+	return nil
+}
+
+func (p *protocol) Stop(_ *manager.Manager) {
+	if p.eventsConsumer != nil {
+		p.eventsConsumer.Stop()
+	}
+}
+
+func (p *protocol) DumpMaps(_ *strings.Builder, _ string, _ *ebpf.Map) {}
```

Comment:
What about the kafkaLastTCPSeqPerConnectionMap map?

Reply:
what's the benefit from that map?